core/src/main/java/io/grpc/util/AdvancedTlsX509TrustManager.java

```diff
@@ -0,0 +1,377 @@
+      return new X509Certificate[0];
+    }
+    return this.delegateManager.getAcceptedIssuers();
+  }
+
+  /**
+   * Uses the default trust certificates stored on user's local system.
+   * After this is used, functions that will provide new credential
+   * data(e.g. updateTrustCredentials(), updateTrustCredentialsFromFile()) should not be called.
+   */
+  public void useSystemDefaultTrustCerts() throws CertificateException, KeyStoreException,
+      NoSuchAlgorithmException {
+    // Passing a null value of KeyStore would make {@code TrustManagerFactory} attempt to use
+    // system-default trust CA certs.
+    X509ExtendedTrustManager newDelegateManager = createDelegateTrustManager(null);
+    this.delegateManager = newDelegateManager;
+  }
+
+  /**
+   * Updates the current cached trust certificates as well as the key store.
+   *
+   * @param trustCerts the trust certificates that are going to be used
+   */
+  public void updateTrustCredentials(X509Certificate[] trustCerts) throws CertificateException,
+      KeyStoreException, NoSuchAlgorithmException, IOException {
+    KeyStore keyStore = KeyStore.getInstance(KeyStore.getDefaultType());
+    keyStore.load(null, null);
+    int i = 1;
+    for (X509Certificate cert: trustCerts) {
+      String alias = Integer.toString(i);
+      keyStore.setCertificateEntry(alias, cert);
+      i++;
+    }
+    X509ExtendedTrustManager newDelegateManager = createDelegateTrustManager(keyStore);
+    this.delegateManager = newDelegateManager;
+  }
+
+  private static X509ExtendedTrustManager createDelegateTrustManager(KeyStore keyStore)
+      throws CertificateException, KeyStoreException, NoSuchAlgorithmException {
+    TrustManagerFactory tmf = TrustManagerFactory.getInstance(
+        TrustManagerFactory.getDefaultAlgorithm());
+    tmf.init(keyStore);
+    X509ExtendedTrustManager delegateManager = null;
+    TrustManager[] tms = tmf.getTrustManagers();
+    // Iterate over the returned trust managers, looking for an instance of X509TrustManager.
+    // If found, use that as the delegate trust manager.
+    for (int j = 0; j < tms.length; j++) {
+      if (tms[j] instanceof X509ExtendedTrustManager) {
+        delegateManager = (X509ExtendedTrustManager) tms[j];
+        break;
+      }
+    }
+    if (delegateManager == null) {
+      throw new CertificateException(
+          "Instance delegateX509TrustManager is null. Failed to initialize");
+    }
+    return delegateManager;
+  }
+
+  private void checkTrusted(X509Certificate[] chain, String authType, SSLEngine sslEngine,
+      Socket socket, boolean checkingServer) throws CertificateException {
+    if (this.verification != Verification.InsecurelySkipAllVerification) {
+      if (chain == null || chain.length == 0) {
+        throw new CertificateException(
+            "Want certificate verification but got null or empty certificates");
+      }
+      if (this.delegateManager == null) {
```

Comment:
Since delegateManager is a volatile, I suggest you store it in a local variable so that this check and the later check{Client,Server}Trusted calls use the same value.